Migrate to `syncOperations` in more places

FranzBusch opened the pull request:
# Motivation

Adding a handler via the regular `pipeline.addHandler` APIs can lead to transferring the handler over isolation regions. To avoid running into `Sendable` warnings here we can use the `syncOperations` of the pipeline to avoid transferring the handlers.

# Modification

This PR is mostly changing code to avoid transferring handlers to and from the event loops in tests. This should be a no-op mostly.

FranzBusch commented on Sources/NIOCore/ChannelPipeline.swift:
```diff
@@ -1082,6 +1082,23 @@ extension ChannelPipeline {
+        ///     - handler: the `ChannelHandler` to remove.
+        /// - returns: the `EventLoopFuture` which will be notified once the `ChannelHandler` was removed.
+        @preconcurrency
+        public func removeHandler(_ handler: RemovableChannelHandler) -> EventLoopFuture<Void> {
```
This is the most relevant change in this PR since it adds a new `removeHandler` method to the `syncOperations`. Importantly this still returns a future since removing is asynchronous but it requires the call to  this method to be from the event loop of the pipeline.

gjcairo commented on Tests/NIOPosixTests/BootstrapTest.swift:
```diff
@@ -448,7 +448,7 @@ class BootstrapTest: XCTestCase {
 
         // But now, it should be there.
-        XCTAssertNoThrow(_ = try server.pipeline.context(name: "AcceptHandler").wait())
+        XCTAssertNoThrow(_ = try server.pipeline.context(name: "AcceptHandler").map { _ in () }.wait())
```
Why is this `map` needed?

FranzBusch replied:
Because the handler is not `Sendable` so we get a warning when we `wait` for it. So mapping to Void avoids that :)

gjcairo reviewed (approved):
LGTM, just left a small question

Lukasa commented on Sources/NIOAsyncAwaitDemo/main.swift:
```diff
@@ -18,10 +18,16 @@ import Dispatch
+                try channel.pipeline.syncOperations.addHTTPClientHandlers()
+                try channel.pipeline.syncOperations.addHandler(NIOHTTPClientResponseAggregator(maxContentLength: 1_000_000))
+                try channel.pipeline.syncOperations.addHandler(MakeFullRequestHandler())
+            }
```
The closure here might well allocate. Probably more complex than just returning a succeeded void future.

Lukasa commented on Sources/NIOChatServer/main.swift:
```diff
@@ -125,9 +125,9 @@ let bootstrap = ServerBootstrap(group: group)
-        channel.pipeline.addHandler(ByteToMessageHandler(LineDelimiterCodec())).flatMap { v in
-            // It's important we use the same handler for all accepted channels. The ChatHandler is thread-safe!
-            channel.pipeline.addHandler(chatHandler)
+        channel.eventLoop.makeCompletedFuture {
```
Same note here.

Lukasa commented on Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift:
```diff
@@ -456,6 +456,29 @@ extension EventLoopFuture {
+        /// - returns: The value of the `EventLoopFuture` when it completes.
+        /// - throws: The error value of the `EventLoopFuture` if it errors.
+        @available(*, noasync, message: "wait() can block indefinitely, prefer get()", renamed: "get()")
+        @preconcurrency
```
I don't think we want `@preconcurrency` here.

Lukasa commented on Sources/NIOCrashTester/OutputGrepper.swift:
```diff
@@ -30,8 +30,10 @@ internal struct OutputGrepper {
             .channelInitializer { channel in
-                channel.pipeline.addHandlers([ByteToMessageHandler(NewlineFramer()),
-                                              GrepHandler(promise: outputPromise)])
+                channel.eventLoop.makeCompletedFuture {
```
Same note here on makeCompletedFuture.

Lukasa commented on Sources/NIOEchoServer/main.swift:
```diff
@@ -46,8 +46,9 @@ let bootstrap = ServerBootstrap(group: group)
         // Ensure we don't read faster than we can write by adding the BackPressureHandler into the pipeline.
-        channel.pipeline.addHandler(BackPressureHandler()).flatMap { v in
-            channel.pipeline.addHandler(EchoHandler())
+        channel.eventLoop.makeCompletedFuture {
```
Same note here on makeCompletedFuture.

Lukasa commented on Sources/NIOPerformanceTester/TCPThroughputBenchmark.swift:
```diff
@@ -119,8 +119,9 @@ final class TCPThroughputBenchmark: Benchmark {
             .channelInitializer { channel in
-                channel.pipeline.addHandler(ByteToMessageHandler(StreamDecoder())).flatMap { _ in
-                    channel.pipeline.addHandler(ClientHandler())
+                channel.eventLoop.makeCompletedFuture {
```
Same note here on `makeCompletedFuture`.

Lukasa commented on Tests/NIOHTTP1Tests/HTTPServerUpgradeTests.swift:
```diff
@@ -38,7 +40,7 @@ extension ChannelPipeline {
         #if !canImport(Darwin) || swift(>=5.10)
         do {
-            _ = try self.context(handlerType: NIOTypedHTTPServerUpgradeHandler<Bool>.self).wait()
+            _ = try self.context(handlerType: NIOTypedHTTPServerUpgradeHandler<Bool>.self).map { _ in () }.wait()
```
This pattern is common enough that you might get more mileage to by factoring this out to a `ChannelPipeline.contains` function.

Lukasa commented on Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift:
```diff
@@ -456,6 +456,29 @@ extension EventLoopFuture {
+        @available(*, noasync, message: "wait() can block indefinitely, prefer get()", renamed: "get()")
+        @preconcurrency
+        @inlinable
+        public func wait(file: StaticString = #file, line: UInt = #line) throws -> Value {
```
This shouldn't be public.

Lukasa replied:
Also, this method is a bit of a code smell to me. In almost all cases this is a deterministic crash, because the only place you are _not_ allowed to call `wait` from is the event loop to which the future belongs. The only time it's safe is when you're on `EmbeddedEventLoop`.

A better fix moves this to a helper method on `EmbeddedEventLoop`.

Lukasa commented on Tests/NIOPosixTests/IdleStateHandlerTest.swift:
```diff
@@ -79,8 +79,9 @@ class IdleStateHandlerTest : XCTestCase {
             .childChannelInitializer { channel in
-                channel.pipeline.addHandler(handler).flatMap { f in
-                    channel.pipeline.addHandler(TestWriteHandler(writeToChannel, assertEventFn))
+                channel.eventLoop.makeCompletedFuture {
```
Same makeCompletedFuture note here.

FranzBusch commented:
@Lukasa Addressed all your comments except the `makeCompletedFuture` once since the work around here is either a manual `do/catch` block or we can overload the `makeCompletedFuture` method for the `Void` case to optimise the allocation to only happen in the error path.

FranzBusch commented:
@Lukasa I updated the allocation numbers for the nighty. Those have been wrong since the nightly was crashing for a long time, hence, I didn't update them last time. The breaking API is expected and not really breaking from what I can see. Can you merge this PR once the CI has passed?

```
10:56:30   💔 API breakage: func ChannelPipeline.addHandler(_:name:position:) is now with @preconcurrency
10:56:30   💔 API breakage: func ChannelPipeline.addHandler(_:name:position:) has parameter 0 type change from NIOCore.ChannelHandler to NIOCore.ChannelHandler & Swift.Sendable
10:56:30   💔 API breakage: func ChannelPipeline.removeHandler(_:) is now with @preconcurrency
10:56:30   💔 API breakage: func ChannelPipeline.removeHandler(_:) has parameter 0 type change from NIOCore.RemovableChannelHandler to NIOCore.RemovableChannelHandler & Swift.Sendable
10:56:30   💔 API breakage: func ChannelPipeline.context(handler:) is now with @preconcurrency
10:56:30   💔 API breakage: func ChannelPipeline.context(handler:) has parameter 0 type change from NIOCore.ChannelHandler to NIOCore.ChannelHandler & Swift.Sendable
10:56:30   💔 API breakage: func ChannelPipeline.addHandlers(_:position:) is now with @preconcurrency
10:56:30   💔 API breakage: func ChannelPipeline.addHandlers(_:position:) has parameter 0 type change from [NIOCore.ChannelHandler] to [NIOCore.ChannelHandler & Swift.Sendable]
10:56:30   💔 API breakage: func ChannelPipeline.addHandlers(_:position:) is now with @preconcurrency
10:56:30   💔 API breakage: func ChannelPipeline.addHandlers(_:position:) has parameter 0 type change from NIOCore.ChannelHandler... to (NIOCore.ChannelHandler & Swift.Sendable)...
```
